### PR TITLE
Implements the AllReduce algorithm.

### PR DESCRIPTION
**Pull request by mbianchif**
Implementa el `WorkerRingManager` y el `AllReduceWorker`.

Metí en este PR también los optimizadores que faltaban porque `AllReduce` si puede usar optimizadores statefull sin estar pasando el estado en cada mensaje.

También aproveché a cambiarle los tipos a los optimizadores porque por ejemplo no sabía que el momentum es un número entre 0 y 1. Lo mismo hice para adam con el nuevo `FloatPositive` y los learning rates. Quedaría agregar todas las variantes a python nomas.

Ya con esto tendríamos todo listo para arrancar con fault tolerance de los nodos.

Se podría ver que onda con el server, por qué es tan superior este método en performance al otro? Capaz se puede optimizar algo en el servidor para que sea más rápido.

Detalle para parameter server, los workers siempre usan descenso por gradiente para calcular el gradiente, es después el server el que sí usa otro tipo de optimizador como adam, descenso por gradiente con momentum, etc, eso no estaba explícito en el adapter, así que lo agregué también.

# on all-reduce
### from the most straightforward algorithm (~parameter server)
1. Every GPU computes the gradient on its subset of the minibatch, and
2. sends it to a single GPU, which in turn
3. averages the gradients that it receives from all other GPUs
4. and sends it back to all of them.
*The more data that's to be shared, the more time it takes to send it*. *NOTE*: another problem I see here is that the single GPU cannot send the averaged gradient in parallel to all other GPUs, so the more of them there are in the system, the more that coordinator spends sending the averaged gradient back. Also, the channel through which the coordinator receives all the gradients, is a single channel, thus, the more GPUs the more time it takes to pull out *all* the bytes of data from it.
#### an example from [[Andrew Gibiansky. Bringing HPC Techniques to Deep Learning|[a paper](https://github.com/lminervino18/oxidized-neural-orchestra/pull/Andrew%20Gibiansky.%20Bringing%20HPC%20Techniques%20to%20Deep%20Learning)]]
Suppose a model with 3 billion parameters, with four bytes each, that's roughly 1.2 gigabytes of data. If the above strategy is used with two GPUs, then each iterations is getting delayed for roughly 1.2 seconds over communication—compare this with the [[jboner. Latency Numbers Every Programmer Should Know|[120 ms RTT CA->Netherlands->CA](https://github.com/lminervino18/oxidized-neural-orchestra/pull/jboner.%20Latency%20Numbers%20Every%20Programmer%20Should%20Know)]].
Parallelizing the training with ten GPUs will slow each iteration by 10.8 seconds—that's 1.2 times 9, so I was right about what I said about the single receiving channel :glasses.

#### scatter-reduce
Each GPU has the complete array of weights, partitioned in $N$ chunks, where $N$ is the amount of GPUs in the ring.
In every iteration, the $n$th GPU will
1. do a backpropagation iteration,
2. send its $n$th chunk of the array,
3. receive the $n-1$th,
4. and aggregate it into its local $n-1$th chunk;
and then proceed backwards from there, sending the chunk it received on each iteration. The algorithm will go on for $N-1$ iterations.
image
By the end, each GPU will contain a chunk that's been aggregated with all other GPUs.
image

#### gather
Same as scatter, except when receiving a chunk, a GPU will overwrite its local one with it, instead of aggregating it.
image
By the end, all GPUs have the final gradient.
image

# References
- [[[Andrew Gibiansky. Bringing HPC Techniques to Deep Learning](https://github.com/lminervino18/oxidized-neural-orchestra/pull/Andrew%20Gibiansky.%20Bringing%20HPC%20Techniques%20to%20Deep%20Learning)]]
- [[[jboner. Latency Numbers Every Programmer Should Know](https://github.com/lminervino18/oxidized-neural-orchestra/pull/jboner.%20Latency%20Numbers%20Every%20Programmer%20Should%20Know)]]

### CODE REVIEW

File: comms/src/connection/connector.rs [hunk 1]
```diff
@@ -54,27 +67,63 @@ where
...
+    /// Connects to a worker and returns a handle to communicate with it.
+    ///
+    /// # Args
+    /// * `id` - The id number of the server.
+    /// * `reader` - The reading end of the communication.
+    /// * `writer` - The writing end of the communication.
+    /// * `src_entity` - The entity initiating the connection.
+    ///
+    /// # Returns
+    /// A [`ParamServerHandle`] ready to start training.
```
Comment (alejoordonez02) [inline comms/src/connection/connector.rs]:
```suggestion
    /// A [`WorkerHandle`] ready to start training.
```

File: comms/src/connection/connector.rs [hunk 2]
```diff
@@ -120,16 +176,17 @@ where
     /// # Args
     /// * `reader` - The reading end of the communication.
     /// * `writer` - The writing end of the communication.
+    /// * `src_entity` - The entity initiating the connection.
     ///
     /// # Returns
     /// A new `ReliableTransport` or an io error if occurred.
```
Comment (alejoordonez02) [inline comms/src/connection/connector.rs]:
```suggestion
    /// A new `TransportLayer` or an io error if occurred.
```

File: comms/src/handles/parameter_server.rs [hunk 3]
```diff
@@ -89,7 +67,7 @@ where
             return Err(io::Error::other(text));
         };
 
-        Ok(PullParamsResponse::Params(params))
+        Ok(params)
```
Comment (alejoordonez02) [inline comms/src/handles/parameter_server.rs:70]: este método quizás o lo pondría en un trait (paja después poner el `get_transport()` igual) o lo movería a un `helper.rs` para q no quede igual en worker y ps

File: comms/src/handles/worker.rs
Comment (alejoordonez02) [inline comms/src/handles/worker.rs:1]: ah hay varios métodos repetidos en los handles, creo que el getter para `transport` no sería tan gede entonces

File: machine_learning/src/optimization/gradient_descent.rs [hunk 4]
```diff
@@ -38,7 +43,7 @@ impl Optimizer for GradientDescent {
         let lr = self.learning_rate;
 
         for (w, g) in params.iter_mut().zip(grad) {
-            *w -= lr * g;
+            *w -= *lr * g;
         }
```
Comment (alejoordonez02) [inline machine_learning/src/optimization/gradient_descent.rs:47]: no digo ahora, pero pensaba que quizás podríamos delegar todas estas cuentas largas a ndarray, tipo sacamos las mut views y hacemos estas cuentas matriciales. Lo bueno de eso es q si lo tenemos todo así después es más fácil hacerlo multithreaded porq es como un switch q tiene el crate

File: machine_learning/src/param_manager.rs [hunk 5]
```diff
@@ -42,15 +67,38 @@ impl<'mw> ParamManager<'mw> {
     ///
     /// # Args
     /// * `servers` - A list of the necessary server metadata to have to iterate through the layers' parameters.
-    /// * `server_ordering` - The ordering of the servers to know which layer corresponds to which server.
+    /// * `layer_ordering` - The ordering of the layers to know which entity holds which layer.
+    ///
+    /// # Returns
+    /// A new `ParamManager` instance.
+    pub fn for_servers(servers: Vec<ParamsMetadata<'mw>>, layer_ordering: &'mw [usize]) -> Self {
```
Comment (alejoordonez02) [inline machine_learning/src/param_manager.rs]: a este le pondría para parameter server

File: machine_learning/src/param_manager.rs [hunk 6]
```diff
@@ -42,15 +67,38 @@ impl<'mw> ParamManager<'mw> {
...
+    /// # Args
+    /// * `params` - The parameters of the worker.
+    /// * `grad` - The gradient buffer of the worker.
+    /// * `residual` - The accumulative gradient buffer of the worker.
+    /// * `amount_of_layers` - The amount of layers in the model.
     ///
     /// # Returns
     /// A new `ParamManager` instance.
-    pub fn new(servers: Vec<ServerParamsMetadata<'mw>>, server_ordering: &'mw [usize]) -> Self {
+    pub fn for_worker(
```
Comment (alejoordonez02) [inline machine_learning/src/param_manager.rs]: y a este para all reduce

File: node/src/main.rs [hunk 7]
```diff
@@ -21,19 +21,35 @@ async fn main() -> io::Result<()> {
...
+    let acceptor = Acceptor::new(transport_factory);
+
+    let transport_factory = |rx, tx| {
+        comms::build_reliable_transport(
+            rx,
+            tx,
+            Duration::from_secs(5),
+            Duration::from_secs(2),
+            2,
+            5,
```
Comment (alejoordonez02) [inline node/src/main.rs]: estos también

File: node/src/main.rs [hunk 8]
```diff
@@ -21,19 +21,35 @@ async fn main() -> io::Result<()> {
...
-        5,
-    );
+        let (rx, tx) = stream.into_split();
+        let transport_layer = comms::build_reliable_transport(
+            rx,
+            tx,
+            Duration::from_secs(5),
+            Duration::from_secs(2),
+            2,
+            5,
```
Comment (alejoordonez02) [inline node/src/main.rs]: estos números los pondría en constantes

File: parameter_server/src/storage/error.rs [hunk 9]
```diff
@@ -3,18 +3,39 @@ use std::{
...
+impl TryInto<ParamServerErr> for MlErr {
+    type Error = ParamServerErr;
+
+    fn try_into(self) -> std::result::Result<ParamServerErr, Self::Error> {
+        match self {
+            Self::SizeMismatch { .. } => Ok(ParamServerErr::SizeMismatch),
+            _ => Err(ParamServerErr::Other),
+        }
+    }
+}
```
Comment (alejoordonez02) [inline parameter_server/src/storage/error.rs:39]: esto no sería mejor `From` o `Into` ? no vi bien dónde se usa pero así no necesitás hacer `??`

File: worker/src/middlewares/worker_ring.rs [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+                }
+                None => chunks[i].fill(0.0),
+            }
+
+            i = (i + n.get() - 1) % n.get();
+            let WorkerEvent::Grad(grad) = self.prev.recv_event().await? else {
+                return Err(io::Error::other("Received an invalid worker event"));
+            };
+
+            for (acc, g) in chunks[i].into_iter().zip(grad) {
```
Comment (alejoordonez02) [inline worker/src/middlewares/worker_ring.rs]: me tira warning

```suggestion
            for (acc, g) in chunks[i].iter_mut().zip(grad) {
```

File: worker/src/middlewares/worker_ring.rs [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+    /// # Returns
+    /// An io error if occurred.
+    async fn scatter(&mut self) -> io::Result<()> {
+        let amount_of_workers = self.addrs.len();
+
+        // SAFETY: In the address list there's at least one
+        //         address, this worker's address.
+        let n = NonZeroUsize::new(amount_of_workers).unwrap();
+        let mut chunks: Vec<_> = self.residual.split_chunks_mut(n).collect();
+        let mut i = self.id;
```
[the remark this answers is not in the document]
Comment (alejoordonez02) [reply worker/src/middlewares/worker_ring.rs]: probé ir guardando el `i` y me da esto

Image

```rs
pub struct WorkerRingManager<T>
where
    T: TransportLayer,
{
    id: usize,
    idx: usize,
// ...

    pub fn new(
        id: usize,
        addrs: Vec<String>,
        prev: WorkerHandle<T>,
        next: WorkerHandle<T>,
        size: usize,
        amount_of_layers: usize,
    ) -> Self {
        Self {
            id,
            idx: id,
// ...

    async fn scatter(&mut self) -> io::Result<()> {
        let amount_of_workers = self.addrs.len();

        // SAFETY: In the address list there's at least one
        //         address, this worker's address.
        let n = NonZeroUsize::new(amount_of_workers).unwrap();
        let mut chunks: Vec<_> = self.residual.split_chunks_mut(n).collect();

        for _ in 0..n.get() - 1 {
            match self.next.push_grad(chunks[self.idx]).await? {
                Some(threshold) => {
                    for g in chunks[self.idx].iter_mut() {
                        if g.abs() >= threshold {
                            *g = 0.0;
                        }
                    }
                }
                None => chunks[self.idx].fill(0.0),
            }

            self.idx = (self.idx + n.get() - 1) % n.get();
            let WorkerEvent::Grad(grad) = self.prev.recv_event().await? else {
                return Err(io::Error::other("Received an invalid worker event"));
            };

            for (acc, g) in chunks[self.idx].iter_mut().zip(grad) {
// ...

    async fn gather(&mut self) -> io::Result<()> {
        let amount_of_workers = self.addrs.len();

        // SAFETY: In the address list there's at least one
        //         address, this worker's address.
        let n = NonZeroUsize::new(amount_of_workers).unwrap();
        let mut chunks: Vec<_> = self.grad.split_chunks_mut(n).collect();
        let mut residual_chunks: Vec<_> = self.residual.split_chunks_mut(n).collect();
        let mut i = (self.idx + 1) % n.get();

        // SAFETY `i` is in bounds, it's defined to be lower than `n`.
        chunks[i].copy_from_slice(residual_chunks[i]);

        if n.get() == 1 {
            residual_chunks[i].fill(0.0);
            return Ok(());
        }

        for j in 0..n.get() - 1 {
            if let Some(threshold) = self.next.push_grad(chunks[i]).await? {
                if j == 0 {
                    for g in residual_chunks[i].iter_mut() {
                        if g.abs() >= threshold {
                            *g = 0.0;
                        }
                    }
                }

                for g in chunks[i].iter_mut() {
                    if g.abs() < threshold {
                        *g = 0.0;
                    }
                }
            } else if j == 0 {
                residual_chunks[i].fill(0.0);
            }

            i = (i + n.get() - 1) % n.get();
            self.idx = i;
            let WorkerEvent::Grad(grad) = self.prev.recv_event().await? else {
                return Err(io::Error::other("Received an invalid worker event"));
            };

            for (acc, g) in chunks[i].iter_mut().zip(grad) {
// ...
```

sin cambiar el idx me da así

Image

así q supongo q me estoy equivocando (tengo sueño 🥱)

File: worker/src/middlewares/worker_ring.rs [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+
+        // SAFETY: In the address list there's at least one
+        //         address, this worker's address.
+        let n = NonZeroUsize::new(amount_of_workers).unwrap();
+        let mut chunks: Vec<_> = self.grad.split_chunks_mut(n).collect();
+        let mut residual_chunks: Vec<_> = self.residual.split_chunks_mut(n).collect();
+        let mut i = (self.id + 1) % n.get();
+
+        // SAFETY `i` is in bounds, it's defined to be lower than `n`.
+        chunks[i].copy_from_slice(&residual_chunks[i]);
```
Comment (alejoordonez02) [inline worker/src/middlewares/worker_ring.rs]:
```suggestion
        chunks[i].copy_from_slice(residual_chunks[i]);
```

File: worker/src/middlewares/worker_ring.rs [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+    ///
+    /// # Returns
+    /// A new `ParamManager` instance or an io error if occurred.
+    async fn gather(&mut self) -> io::Result<()> {
+        let amount_of_workers = self.addrs.len();
+
+        // SAFETY: In the address list there's at least one
+        //         address, this worker's address.
+        let n = NonZeroUsize::new(amount_of_workers).unwrap();
+        let mut chunks: Vec<_> = self.grad.split_chunks_mut(n).collect();
```
Comment (alejoordonez02) [inline worker/src/middlewares/worker_ring.rs:153]: estos split chunks no sé si los tendría como un atributo porq no deberían cambiar para toda la corrida de all reduce

File: worker/src/middlewares/worker_ring.rs [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+            }
+
+            i = (i + n.get() - 1) % n.get();
+            let WorkerEvent::Grad(grad) = self.prev.recv_event().await? else {
+                return Err(io::Error::other("Received an invalid worker event"));
+            };
+
+            for (acc, g) in chunks[i].into_iter().zip(grad) {
+                *acc = *g;
+            }
```
Comment (alejoordonez02) [inline worker/src/middlewares/worker_ring.rs]:
```suggestion
            chunks[i].copy_from_slice(grad);
```

**Review by alejoordonez02 — APPROVED** [review]
quedó god, las cosas q moviste muy clean también.
te dejé un par de comentarios boludos,
más que nada quizás sí movería los métodos que están repetidos en los handle a un trait e implementaría los getter q hagan falta para poder tener los métodos ahí. Si fueran menos métodos repetidos hacer los getters sería más paja pero creo q acá queda mejor.
te la dejo aprobada

File: comms/src/handles/worker.rs
Comment (mbianchif) [reply comms/src/handles/worker.rs:1]: Siendo que ya está hecho así y no vamos a hacer más handles lo dejaría como está. No lo vamos a estar usando como un generic en ningún lado, sino siempre como tipos concretos. Capaz cuando hagamos la parte de fault tolerance agreguemos más métodos y sea más evidente pero de momento con que esté así no me parece mal. Prefiero esto a tener el getter del transport.

File: machine_learning/src/optimization/gradient_descent.rs (hunk 4 above)
Comment (mbianchif) [reply machine_learning/src/optimization/gradient_descent.rs:47]: Sí, ahora que están todos los optimizers en `machine_learning` creo que va a quedar bien.

File: worker/src/middlewares/worker_ring.rs (hunk 13 above)
Comment (mbianchif) [reply worker/src/middlewares/worker_ring.rs:153]: No entendí, a que te referís con tener como atributo?
